libr/anal/xrefs.c

```diff
@@ -2,51 +2,134 @@
 
 #include <r_anal.h>
 #include <r_cons.h>
+#include <r_vec.h>
 
-static RAnalRef *r_anal_ref_new(ut64 addr, ut64 at, ut64 type) {
-	RAnalRef *ref = R_NEW (RAnalRef);
-	if (ref) {
-		ref->addr = addr;
-		ref->at = at;
-		ref->type = (type == UT64_MAX)? R_ANAL_REF_TYPE_CODE: type;
+R_GENERATE_VEC_IMPL_FOR(AnalRef, RAnalRef);
+
+typedef struct r_ref_end_t {
+	ut64 to;
+	RAnalRefType type;
+} RefEnd;
+
+R_GENERATE_VEC_IMPL_FOR(RefEnd, RefEnd);
+
+typedef struct r_ref_edges_t {
+	ut64 from;
+	RVecRefEnd tos;
+} RefEdges;
+
+R_GENERATE_VEC_IMPL_FOR(AdjacencyList, RefEdges);
+
+// TODO later turn into struct and add caching (using last used pointers)?
+typedef RVecAdjacencyList AdjacencyList;
+
+typedef struct r_ref_manager_t {
+	AdjacencyList refs;   // forward refs
```

Comment:
isnt this O(n) vs the previous O(1) of the hashtable?

Reply:
With the new index data structure lookup should be close to O(1) again.
Also the current implementation is 35s vs 45s before, so maybe memory locality wins over O(n) in this case?